### ISSUE TITLE
Sign data with key in PEM

### ISSUE THREAD

**Issue by assafb81** [dvsekhvalnov/jose-jwt#76]
I want to work with keys that are stored in PEM files.
I am using jsrsasign library (part of supported libraries from [jwt.io](https://jwt.io/))

I could not understand how I can in **C#**:
1.read private key, stored in PEM file.
2. sign data

using 'jose-jwt' dll.
I tried to use BouncyCaslte dll but I am getting different key and signature from the one I am getting when I am using jsrsasign in the client side.
I want to use 'SHA256withRSA' alg.

Thanks

**Comment by dvsekhvalnov** [comment]
Hi @assafb81 ,

did you mean `RS256` signing algo? It expects `RSA` type of key to be passed.
The simplest will be to combine your private key with certificate into .p12 file, then it can be loaded natively with `X509Certificate2` 

If you really wanna stick with PEM file, here is article describing how to do it in C#: https://www.codeproject.com/Articles/162194/Certificates-to-DB-and-Back

there is also similar ticket about EC keys, may be you can get some ideas:
https://github.com/dvsekhvalnov/jose-jwt/issues/25

I'm considering adding helper to library to work with popular formats, there were quite a few requests and questions about it recently. But it is not yet in a library.

**Comment by assafb81** [comment]
Hi @dvsekhvalnov,

Thanks for the quick reply.
I want to stick with PEM files because the client side library uses PEM to verify the data sent from the server.
All the links you provided are using CER files.
I look at [issue - 30](https://github.com/dvsekhvalnov/jose-jwt/issues/30) with my keys which are in PEM format that looks something like:
**private**
> -----BEGIN RSA PRIVATE KEY-----
Private KEY
-----END RSA PRIVATE KEY-----

when I sign my data (simple data: data = "Test") I've got different signature values:
client side code:
```
const privKey = rsa.KEYUTIL.getKey(PRIVATE_KEY);
const sig = new rsa.Signature({ alg: 'SHA256withRSA' });
sig.init(privKey);
sig.updateString('Test');
const signature = sig.sign();
```

signature value - `3f6de53a3987203122d1bb380d59b8f2f39b536b0352ae1a6d290206f1c719d59f8426f51990612aab402e720d4d0ab02c98382e845ae485794868a5f7ce53c8bb23575193ea07c266ecae82f29dea3cdaa1619017f38d3f74d4eb14461d896cfb23aed80ebd36420fe098636cce7dda2b6c2c286919d992cffa4e1114867695`

server side code (I even did the round trip to xml and back, don't know why its needed):
```
StreamReader reader = new StreamReader(PrivateKeyPath);
PemReader pemReader = new PemReader(reader);
AsymmetricCipherKeyPair keyPair = (AsymmetricCipherKeyPair)pemReader.ReadObject();
AsymmetricKeyParameter privateKey = keyPair.Private;
RSA rsa = DotNetUtilities.ToRSA((RsaPrivateCrtKeyParameters)privateKey);
string prvKeyXmlString = rsa.ToXmlString(true);

RSACryptoServiceProvider key = new RSACryptoServiceProvider();
key.FromXmlString(prvKeyXmlString);

string dd = "Test";
var dataTosign = Encoding.UTF8.GetBytes(dd);
var signature = key.SignData(dataTosign, CryptoConfig.MapNameToOID("SHA256"));

```
signature value - `P23lOjmHIDEi0bs4DVm48vObU2sDUq4abSkCBvHHGdWfhCb1GZBhKqtALnINTQqwLJg4LoRa5IV5SGil985TyLsjV1GT6gfCZuyugvKd6jzaoWGQF/ONP3TU6xRGHYls+yOu2A69NkIP4JhjbM592itsLChpGdmSz/pOERSGdpU=`

I suspect that in the conversion of the key in the server side and the client side from PEM to keyObject is the issue.

**Comment by dvsekhvalnov** [comment]
@assafb81 , CER and PEM is often used interchangeably. It's really same thing.

Here is SO article : https://stackoverflow.com/questions/7400500/how-to-get-private-key-from-pem-file

though it refers to same CodeProject link :)

**Comment by assafb81** [comment]
@dvsekhvalnov, there is no confirmed solution there.
the answer by "Marnix van Valen" with the `GetBytesFromPEM(string pemString, string section)` method 
can't create new X509Certificate2 object but I will continue to work on that.
getting "Cannot find the requested object" error

**Comment by dvsekhvalnov** [comment]
I see. Well i'd suggest to simply convert your pem to p12 then. It's pretty straightforward with openssl.

**Comment by assafb81** [comment]
got any references for 'PEM to p12 conversion' in C#?

**Comment by dvsekhvalnov** [comment]
You don't need to do it in C#, openssl will work just fine:

openssl pkcs12 -export -out certificate.p12 -inkey privateKey.pem -in certificate.crt 

(you can also try online converters like https://www.sslshopper.com/ssl-converter.html) But be aware, i'd not trust production private keys to anybody.

**Comment by assafb81** [comment]
I need C# because I am writing a program that will do that for me.
I can't use online tool for my purposes.

**Comment by dvsekhvalnov** [comment]
Sorry, didn't get that. Why openssl is not working? Just convert your private key to p12 and that's it.

**Comment by assafb81** [comment]
I have a private key stored in PEM file.
the flow goes like this:
server side/console application: get data -> sign -> save signed data
client side (web/mobile application): fetch data  -> validate signature -> continue according to validation response.

I need to be able to sign data that keeps changing.

do you suggest to convert the PEM to p12 once. and than use p12 for signing?

**Comment by dvsekhvalnov** [comment]
Yep exactly, just convert your private key from PEM to p12. p12 is natively supported by C#.

**Comment by assafb81** [comment]
when I run this command:
'openssl pkcs12 -export -out certificate.p12 -inkey privateKey.pem -in certificate.crt'

I don't have cetificate.crt... how imporant is it?

**Comment by dvsekhvalnov** [comment]
1. Given you have private key you can always generate certificate, something like:

openssl req -new -x509 -days 1826 -key privateKey.pem -out certificate.crt

2. sha256 - is hash algorithm, has nothing to do with .p12 file

**Comment by assafb81** [comment]
the PEM file should contain the private key and the public key in order to created the certification?

**Comment by dvsekhvalnov** [comment]
1. Public key is just part of private, you need only private.

2. A lot in docs: https://github.com/dvsekhvalnov/jose-jwt#rs--and-ps--family

**Comment by assafb81** [comment]
my code is:

```
var payload = new Dictionary<string, object>()
{
    {"sub", data}
};

var privateKey = new X509Certificate2("../../Keys/certificate.p12", "password", X509KeyStorageFlags.Exportable | X509KeyStorageFlags.MachineKeySet).PrivateKey as RSACryptoServiceProvider;
string token = Jose.JWT.Encode(payload, privateKey, JwsAlgorithm.RS256);
```

and I am getting in 'new X509....' line the error:

System.Security.Cryptography.CryptographicException was unhandled
  HResult=-2147024810
  Message=The specified network password is not correct.

  Source=mscorlib
  StackTrace:
       at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)
       at System.Security.Cryptography.X509Certificates.X509Utils._LoadCertFromFile(String fileName, IntPtr password, UInt32 dwFlags, Boolean persistKeySet, SafeCertContextHandle& pCertCtx)
       at System.Security.Cryptography.X509Certificates.X509Certificate.LoadCertificateFromFile(String fileName, Object password, X509KeyStorageFlags keyStorageFlags)
       at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(String fileName, String password, X509KeyStorageFlags keyStorageFlags)
       at RegimenSignature.Program.SignRegimen(String path) in C:\LocalProjects\Project\trunk\RegimenSignature\RegimenSignature\Program.cs:line 116
       at RegimenSignature.Program.Main(String[] args) in C:\LocalProjects\Project\trunk\RegimenSignature\RegimenSignature\Program.cs:line 64
       at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
  InnerException: 


1. I am using my data as the subject in the payload.
2.  what is the "password" (the 2nd parameter of the function) should be?

**Comment by dvsekhvalnov** [comment]
This is password to your .p12 file. openssl have asked you for password when you were converting.

**Comment by assafb81** [comment]
I changed the password to .p12 file and it passes the line.
now it is stuck on the next on:
`string token = Jose.JWT.Encode(payload, privateKey, JwsAlgorithm.RS256);`

the exception:
Invalid algorithm specified

**Comment by dvsekhvalnov** [comment]
Can you post full stack trace?

**Comment by assafb81** [comment]
System.Security.Cryptography.CryptographicException was unhandled
  HResult=-2146893816
  Message=Invalid algorithm specified.

  Source=mscorlib
  StackTrace:
       at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)
       at System.Security.Cryptography.Utils.SignValue(SafeKeyHandle hKey, Int32 keyNumber, Int32 calgKey, Int32 calgHash, Byte[] hash, Int32 cbHash, ObjectHandleOnStack retSignature)
       at System.Security.Cryptography.Utils.SignValue(SafeKeyHandle hKey, Int32 keyNumber, Int32 calgKey, Int32 calgHash, Byte[] hash)
       at System.Security.Cryptography.RSACryptoServiceProvider.SignHash(Byte[] rgbHash, Int32 calgHash)
       at System.Security.Cryptography.RSAPKCS1SignatureFormatter.CreateSignature(Byte[] rgbHash)
       at Jose.RsaUsingSha.Sign(Byte[] securedInput, Object key)
       at Jose.JWT.EncodeBytes(Byte[] payload, Object key, JwsAlgorithm algorithm, IDictionary`2 extraHeaders, JwtSettings settings)
       at Jose.JWT.Encode(String payload, Object key, JwsAlgorithm algorithm, IDictionary`2 extraHeaders, JwtSettings settings)
       at Jose.JWT.Encode(Object payload, Object key, JwsAlgorithm algorithm, IDictionary`2 extraHeaders, JwtSettings settings)
       at RegimenSignature.Program.SignRegimen(String path) in C:\LocalProjects\Project\trunk\RegimenSignature\RegimenSignature\Program.cs:line 118
       at RegimenSignature.Program.Main(String[] args) in C:\LocalProjects\Project\trunk\RegimenSignature\RegimenSignature\Program.cs:line 64
       at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
  InnerException:

**Comment by dvsekhvalnov** [comment]
It is in the docs: https://github.com/dvsekhvalnov/jose-jwt#rsacryptoserviceprovider

**Comment by assafb81** [comment]
Thanks for your HELP!!!!!!!
I used [create jwt with a private rsa key](http://www.donaldsbaconbytes.com/2016/08/create-jwt-with-a-private-rsa-key/) from the link above and finally after a long long time it is working.
I manage to sign on the server with private key stored in PEM file and verify on the client side using JSRSASIGN package.

**Comment by assafb81** [comment]
I manage to create a token in the server side (using [create jwt with a private rsa key](http://www.donaldsbaconbytes.com/2016/08/create-jwt-with-a-private-rsa-key/)) and validate it on the client side using JSRSASIGN package.

**Comment by dvsekhvalnov** [comment]
Glad it working @assafb81 :)

Feel free to close ticket if no questions left.

**Comment by assafb81** [comment]
Closes #76

**Comment by assafb81** [comment]
Close #76
